From jcb@inf.ed.ac.uk  Wed Mar  1 15:31:40 2006
Message-ID: <17413.30016.470872.240977@palau.inf.ed.ac.uk>
Date: Wed, 1 Mar 2006 10:19:44 +0000
From: Julian Bradfield <jcb+luu@inf.ed.ac.uk>
cc: Greg KH <greg@kroah.com>
Subject: USB: PL2303 and TIOCMIWAIT

A while ago, I posted about TIOCMIWAIT not working with the PL2303
USB-serial adapter.


After a brief exchange with Greg, I tracked this to a missing wake-up
in the USB interrupt procedures. I got our systems staff to install
the enclosed very simple patch to our 2.6.12 kernels, and it all works
fine as expected. I guess this should also apply to the latest version
and go into the mainstream.

Apologies for the long delay in posting the result.

The routine being patched is pl2303_update_line_status

Signed-off-by: Julian Bradfield <jcb+luu@inf.ed.ac.uk>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


---
 drivers/usb/serial/pl2303.c |    1 +
 1 file changed, 1 insertion(+)

--- gregkh-2.6.orig/drivers/usb/serial/pl2303.c
+++ gregkh-2.6/drivers/usb/serial/pl2303.c
@@ -826,6 +826,7 @@ static void pl2303_update_line_status(st
 	spin_lock_irqsave(&priv->lock, flags);
 	priv->line_status = data[status_idx];
 	spin_unlock_irqrestore(&priv->lock, flags);
+	wake_up_interruptible (&priv->delta_msr_wait);
 
 exit:
 	return;
